### PR TITLE
#163519140 add reset password via email endpoint

### PR DESCRIPTION
**Pull request by mcaleb808**
#### What does this PR do?
adds reset the password via email endpoint

#### Description of Task to be completed?
- forget password endpoint
- send reset password email to a user
- reset password endpoint
- send a confirmation email to a user
- unit tests

#### How should this be manually tested?
- git clone this branch

- run `npm install`

- run `npm migrate:seed`

- run `npm run dev`

- with postman use this link to signup:
`POST: api/v1/users/forget`

- Body request example (forget email):
```request-json
{
   "email":"test@mail.com"
}
```
- After you check email with reset password link and copy it or click it
-Go back to postman with PUT method copy the link in URL
`PUT: api/v1/users//:userId/reset/:resetCode`

- Body request example (reset password):
```request-json
{
   "newPassword":"password",
    "confirmNewPassword":"password"
}
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
`#163519140`

#### Screenshots (if appropriate)
- Reset password email
image

- Reset password confirmation email
image

#### Questions:
N/A

### CODE REVIEW

File: __tests__/routes/user.test.js [hunk 1]
```diff
@@ -51,4 +51,73 @@ describe('users', () => {
     expect(res.status).toBe(401);
     expect(res.body.message).toBe('test@email.com has already been confirmed');
   });
+
+  // Forget password tests
```
Comment (kimotho-njoki) [inline __tests__/routes/user.test.js]: Like I said. One line comments are unnecessary. Please delete them. I can understand this is a `password reset` test without the comment.

File: __tests__/routes/user.test.js [hunk 2]
```diff
@@ -51,4 +51,73 @@ describe('users', () => {
...
+  }, 30000);
+
+  test('No user found with that email address', async () => {
+    const res = await request(app)
+      .post(`${urlPrefix}/users/forget`)
+      .send({ user: { email: 'fake@email.com' } });
+    expect(res.status).toBe(404);
+    expect(res.body.message).toBe('No user found with that email address');
+  });
+  test('Bad request- password forget', async () => {
```
Comment (kimotho-njoki) [inline __tests__/routes/user.test.js:78]: Empty line after line 70. Why have you spaced the rest of the tests but not this one?

File: __tests__/routes/user.test.js (hunk 1 above)
Comment (mcaleb808) [reply __tests__/routes/user.test.js]: Right away!

File: __tests__/routes/user.test.js (hunk 2 above)
Comment (mcaleb808) [reply __tests__/routes/user.test.js:78]: It was unintentionally! let me change that too

File: controllers/AuthController.js [hunk 3]
```diff
@@ -78,5 +78,94 @@ class AuthController {
...
+      const createReset = await ResetPassword.create({ userId: id });
+      const { resetCode } = createReset.get();
+      await resetPasswordEmail(id, email, resetCode);
+      res.status(201).json({
+        status: 201,
+        message: 'Password reset link sent sucessfully. Please check your email!',
+        ResetPassword: createReset
+      });
+    } catch (error) {
+      return res.status(401).json({ message: 'Please try again' });
```
Comment (Nziranziza) [inline controllers/AuthController.js]: I suggest this status code to be something else not 401 for example 520 for unknown error

File: controllers/AuthController.js [hunk 4]
```diff
@@ -78,5 +78,94 @@ class AuthController {
...
+          await newPasswordEmail(user.email);
+          res.status(200).json({
+            status: 200,
+            user: { ...user.get() },
+            message: 'Your password has been reset successfully!'
+          });
+        }
+      }
+    } catch (error) {
+      return res.status(401);
```
Comment (Nziranziza) [inline controllers/AuthController.js]: Here also you can change the status

**Review by Nziranziza — APPROVED** [review]

File: controllers/AuthController.js (hunk 3 above)
Comment (mcaleb808) [reply controllers/AuthController.js]: I think it is a good idea!

**Pull request merged**